src/Symfony/Component/Validator/GroupProviderFactory.php

```diff
@@ -0,0 +1,29 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Validator;
+
+/**
+ * Factory class for creating or retrieving instances of validation group providers.
+ *
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ */
+class GroupProviderFactory implements GroupProviderFactoryInterface
+{
+    public function __construct(private array $providers = [])
+    {
+    }
+
+    public function getInstance(string $class): GroupProviderInterface
+    {
+        return $this->providers[$class] ??= new $class();
```

Comment:
Can't we get rid of this class + interface and rely on ContainerInterface instead.
Also on this line, I'd prefer not doing any `new` and instead always require a service.
This would remove the requirement for the key to be a class name, which would be better IMHO (allows targetting a different service of the same class if needed.)

Reply:
>Can't we get rid of this class + interface and rely on ContainerInterface instead?

I'm not sure to understand your suggestion 😕 here 

>Also on this line, I'd prefer not doing any new and instead always require a service.
This would remove the requirement for the key to be a class name, which would be better IMHO (allows targetting a different service of the same class if needed.)

I believe that requirement exists only when the component is used alone. I copied that strategy from [`ConstraintValidatorFactory`](https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Validator/ConstraintValidatorFactory.php), which uses the same approach for class instantiation.

On the other hand, when the component is used in conjunction with the Framework, it's the `ContainerGroupProviderFactory` that will be used by default, allowing targeting of a different service of the same class if needed. Same approach as `ContainerConstraintValidatorFactory`.

Reply:
I mean: `GroupProviderInterface` defines a service locator, so let's use ContainerInterface instead?
For standalond usage, it's still quite easy to build a locator.

Reply:
I'm delving into it again, but I don't see how we can reference `ContainerInterface` when the component is used outside the framework stack, especially since it's installed (transitively) as a dev requirement, so I'm confuse, could you please elaborate a bit more or provide an example?

Btw, as it stands, this class allows targeting a different service of the same class if desired:
```php
$factory = new GroupProviderFactory([
    'key1' => new MyGroupProvidor($groups1),
    'key2' => new MyGroupProvidor($groups2),
])
```